app/models/host/managed.rb

```diff
@@ -551,11 +551,33 @@ def jumpstart?
     operatingsystem.family == "Solaris" and architecture.name =~/Sparc/i rescue false
   end
 
+  def hostgroup_inherited_attributes
+    %w{puppet_proxy_id puppet_ca_proxy_id environment_id compute_profile_id}
+  end
+
+  def apply_inherited_attributes(attributes)
+    return nil unless attributes
+    new_hostgroup_id = attributes['hostgroup_id']
+    #hostgroup didn't change, no inheritance needs update.
+    return attributes if new_hostgroup_id.blank?
+
+    hostgroup = self.hostgroup || Hostgroup.find_by_id(new_hostgroup_id)
+    hostgroup = Hostgroup.find_by_id(new_hostgroup_id) unless hostgroup.id == new_hostgroup_id
+    return attributes unless hostgroup
+
+    inherited_attributes = hostgroup_inherited_attributes - attributes.keys
+
+    inherited_attributes.each do |attribute|
+      value = hostgroup.send("inherited_#{attribute}")
+      attributes[attribute] = value
+    end
+
+    attributes
+  end
+
   def set_hostgroup_defaults
     return unless hostgroup
-    assign_hostgroup_attributes(%w{environment_id domain_id compute_profile_id})
-    assign_hostgroup_attributes(["puppet_proxy_id"]) if new_record? || (!new_record? && !puppet_proxy_id.blank?)
-    assign_hostgroup_attributes(["puppet_ca_proxy_id"]) if new_record? || (!new_record? && !puppet_ca_proxy_id.blank?)
+    assign_hostgroup_attributes(%w{domain_id compute_profile_id})
```

Comment:
Why only change the behaviour of some attributes, shouldn't they all work this way?